app/lib/search/index_ducene.dart

```diff
@@ -0,0 +1,226 @@
+    await _init();
+    await _index.deleteDocuments(new BoolQuery()..append('id', url));
+  }
+
+  @override
+  Future merge() async {
+    await _init();
+    await _index.forceMerge();
+    _lastUpdated = new DateTime.now();
+  }
+
+  @override
+  Future<PackageSearchResult> search(PackageQuery query) async {
+    if (!isReady) return new PackageSearchResult.notReady();
+
+    await _init();
+    final IndexSearcher searcher = _index.newRealTimeIndexSearcher();
+    final String queryText = _normalizeText(query.text, 200);
+    final duceneQuery = new BoolQuery()
+      ..append('package', queryText, analyzer: _analyzer, boost: 4.0)
+      ..append('description', queryText, analyzer: _analyzer, boost: 2.0)
+      ..append('readme', queryText, analyzer: _analyzer);
+    if (query.type != null) {
+      duceneQuery
+          .addFilter(new TermQuery(new Term('detectedTypes', query.type)));
+    }
+
+    final int offset = query.offset ?? 0;
+    final int limit = max(minSearchLimit, query.limit ?? 100);
+
+    final TopDocs topPackages = await searcher.search(
+      duceneQuery,
+      maxSearchResults + minSearchLimit, // additional results for extended bias
+      scoreSort: new TFIDFScoreSort(),
+    );
+
+    List<PackageScore> packages = [];
+    for (ScoreDoc sd in topPackages.scoreDocs) {
+      final Document doc = await searcher.doc(sd.doc);
+      final double bias = double.parse(doc.get('popularity')) * 20;
+
+      packages.add(new PackageScore(
+        doc: new PackageDocument(
+          url: doc.get('id'),
+          package: doc.get('package'),
+          version: doc.get('version'),
+          devVersion: doc.get('devVersion'),
+          description: doc.get('originalDescription'),
+          lastUpdated: doc.get('lastUpdated'),
+          detectedTypes: doc.getValues('detectedTypes'),
+        ),
+        score: sd.score + bias,
+      ));
+    }
+    packages.sort((a, b) => -Comparable.compare(a.score, b.score));
+    if (packages.length > maxSearchResults) {
+      packages = packages.sublist(0, maxSearchResults);
+    }
+    final int totalCount = packages.length;
+
+    if (packages.length <= offset) {
+      packages = [];
+    } else {
+      packages = packages.sublist(offset);
+    }
+    if (packages.length > limit) {
+      packages = packages.sublist(0, limit);
+    }
+
+    return new PackageSearchResult(
+      lastUpdated: _lastUpdated.toIso8601String(),
+      totalCount: totalCount,
+      packages: packages,
+    );
+  }
+
+  String _normalizeWhitespaces(String text, int maxLength) {
+    if (text == null) return '';
+    String t = text.replaceAll(new RegExp('\\s+'), ' ').trim();
+    if (t.length > maxLength) {
+      t = t.substring(0, maxLength);
+    }
+    return t;
+  }
+
+  String _normalizeText(String text, int maxLength) {
+    if (text == null) return '';
+    String t = text
+        .toLowerCase()
+        .replaceAll(new RegExp('\\W'), ' ')
+        .replaceAll(new RegExp('\\s+'), ' ')
```

Comment:
Make these global final RegExp's so the VM can optimize them properly when invoking many times.